File: cs259/project/matlab/bandpassIIR.c
// BANDPASS IIR FILTER
// bandpass_output: bandpass filtered output
// input_data: input samples to be filtered
// input_size: number of input samples
// num_coeffs: IIR numerator filter coefficients
// denom_coeffs: IIR denominator filter coefficients
// n_coeffs: number of filter coefficients

void lowpassFIR(float *lowpass_output, float *input_data, int input_size, float *num_coeffs, float *denom_coeffs, int n_coeffs) {
    // different equation representation of bandpass IIR filter
    int i, k, k_limit;
    for (i=0; i<input_size; i++) {
        k_limit = (k < n_coeffs) ? k : n_coeffs;
        for (k=0; k<k_limit; k++) {
            lowpass_output[i] += -denom_coeffs[k]*lowpass_output[i-k] + num_coeffs[k]*input_data[i-k];
        }
    }

    return;
}

File: digitalFiltersDesigns/bandpassIIR.m
function [ bandpass_output ] = bandpassIIR( input_data, num_filter_coeffs, denom_filter_coeffs )
%bandpassIIR -- IIR butterworth bandpass filter
%   bandpass filters the input data using the filter coeffcients provided
%   as arguments to the function.
% num_filter_coeffs are the filter coeffcients of numerator
% denom_filter_coeffs are the filter coefficients of the denominator
N_samples = length(input_data); % no. samples in input data
N_coeffs  = length(num_filter_coeffs); % no. filter coefficients

bandpass_output = zeros(1, N_samples);


for i=1:N_samples
   for k=1:min(i, N_coeffs)
        bandpass_output(i) = bandpass_output(i) - ...
            denom_filter_coeffs(k)*bandpass_output(i-k+1) ...
            + num_filter_coeffs(k)*input_data(i-k+1);
   end
end

end

